Sources/TensorFlow/Operators/Basic.swift

```diff
@@ -390,44 +390,92 @@ public extension Tensor {
     ///
     /// - Parameters:
     ///   - indices: Contains the indices to gather.
+    ///   - axis: Dimension along which to gather. Negative values wrap around.
     ///   - batchDims: Number of leading batch dimensions to ignore.
     ///
+    /// - Precondition: `axis` must be in the range `[-rank, rank)`, while also being greater than
+    ///   or equal to `batchDims`.
     /// - Precondition: `batchDims` must be less than `indices.rank`.
     ///
     /// - Returns: The gathered tensor.
     @inlinable
     @differentiable(wrt: self where Scalar: TensorFlowFloatingPoint)
-    func batchGathering<Index: TensorFlowIndex>(atIndices indices: Tensor<Index>) -> Tensor {
-        var batchIndices = indices
-        var accumulated = Tensor<Index>(ones: [])
-        accumulated *= withoutDerivative(at: shapeTensor) { Tensor<Index>($0[1]) }
-        let dValue = withoutDerivative(at: shapeTensor) { $0[0] }
-        let dIndices = Tensor<Index>(
-            rangeFrom: Tensor<Index>(zeros: []),
-            to: Tensor<Index>(dValue),
-            stride: Tensor<Index>(ones: [])
-        ) * accumulated
-        let dShape = Tensor<Int32>(concatenating: [
-            dValue.rankLifted(),
-            Tensor<Int32>([Int32](repeating: 1, count: indices.rank - 1))])
-        batchIndices += dIndices.reshaped(toShape: dShape)
+    func batchGathering<Index: TensorFlowIndex>(
+        atIndices indices: Tensor<Index>,
+        alongAxis axis: Int = 1,
+        batchDimCount batchDims: Int = 1
+    ) -> Tensor {
+        // TODO: precondition(batchDims >= 0 && batchDims < indices.rank,
+        //                    "'numBatchDims' must be non-negative and less than 'indices.rank'.")
+        // TODO: precondition(batchDims < rank, 
+        //                    "'numBatchDims' must be less than the tensor's rank.")
+
+        // Handle the axis argument by transposing the axis dimension so that it is the first
+        // non-batch dimension, recursively calling `batchGathering` with `axis = 0`, and then
+        // transposing the result to put the pre-axis dimensions before the indices dimensions.
+        if axis != batchDims {
+            // Adjust axis to be positive.
+            let posAxis = axis < 0 ? axis + rank : axis
+
+            // TODO: precondition(posAxis >= 0 && posAxis < rank, "'axis' is out of range.")
+            // TODO: precondition(batchDims <= posAxis,
+            //                    "'batchDims' must be less than or equal to 'axis'.")
+
+            // Move self[axis] up to self[batchDims].
+            let permutation = Tensor<Int32>(concatenating: [
+                Tensor<Int32>(rangeFrom: 0, to: Int32(batchDims), stride: 1),
+                Tensor<Int32>(Int32(axis)).rankLifted(),
+                Tensor<Int32>(rangeFrom: Int32(batchDims), to: Int32(posAxis), stride: 1),
+                Tensor<Int32>(rangeFrom: Int32(axis) + 1, to: Int32(rank), stride: 1)])
```

Comment:
```suggestion
                Tensor(rangeFrom: Int32(axis) + 1, to: Int32(rank), stride: 1)])
```

Reply:
I cannot remove this from any of the calls to this constructor because removing it results in an error about ambiguity at compile time.

Reply:
Oops. Sorry!